crates/trie/src/proof.rs

```diff
@@ -55,226 +40,196 @@ where
+                    proof.set_value(value);
+                }
+
+                hash_builder
+                    .add_leaf(hashed_key_nibbles, alloy_rlp::encode_fixed_size(&value).as_ref());
+                storage = hashed_storage_cursor.next()?;
             }
         }
 
-        self.node_rlp_buf.clear();
-        BranchNode::new(&branch_node_stack).rlp(node.state_mask, &mut self.node_rlp_buf);
-        Ok(Bytes::copy_from_slice(self.node_rlp_buf.as_slice()))
-    }
-
-    /// Restore leaf node.
-    /// The leaf nodes are always encoded as `RLP(node) or RLP(keccak(RLP(node)))`.
-    fn restore_leaf_node(&mut self, seek_key: B256, slice_at: usize) -> Result<Bytes, ProofError> {
-        let (hashed_address, account) = self
-            .hashed_account_cursor
-            .seek(seek_key)?
-            .ok_or(ProofError::LeafAccountMissing(seek_key))?;
-
-        // Restore account's storage root.
-        let storage_root = StorageRoot::new_hashed(self.tx, hashed_address)
-            .with_hashed_cursor_factory(self.hashed_cursor_factory.clone())
-            .root()?;
-
-        self.account_rlp_buf.clear();
-        EthAccount::from(account).with_storage_root(storage_root).encode(&mut self.account_rlp_buf);
-
-        let leaf_node_key = Nibbles::unpack(hashed_address).slice_from(slice_at);
-        let leaf_node = LeafNode::new(&leaf_node_key, &self.account_rlp_buf);
-
-        self.node_rlp_buf.clear();
-        Ok(Bytes::from(leaf_node.rlp(&mut self.node_rlp_buf)))
-    }
-
-    /// Restore target leaf node.
-    /// The target node has to have an exactly matching key and is always encoded as `RLP(node)`.
-    /// The target node might be missing from the trie.
-    fn restore_target_leaf_node(
-        &mut self,
-        seek_key: B256,
-        slice_at: usize,
-    ) -> Result<Option<Bytes>, ProofError> {
-        let (hashed_address, account) = match self.hashed_account_cursor.seek(seek_key)? {
-            Some(entry) if entry.0 == seek_key => entry,
-            _ => return Ok(None),
-        };
-
-        // Restore account's storage root.
-        let storage_root = StorageRoot::new_hashed(self.tx, hashed_address)
-            .with_hashed_cursor_factory(self.hashed_cursor_factory.clone())
-            .root()?;
-
-        self.account_rlp_buf.clear();
-        EthAccount::from(account).with_storage_root(storage_root).encode(&mut self.account_rlp_buf);
-
-        let leaf_node_key = Nibbles::unpack(hashed_address).slice_from(slice_at);
-        let leaf_node = LeafNode::new(&leaf_node_key, &self.account_rlp_buf);
+        let root = hash_builder.root();
+
+        let proof_nodes = hash_builder.take_proofs();
+        for proof in proofs.iter_mut() {
+            proof.set_proof(
+                proof_nodes
+                    .iter()
+                    .filter(|(path, _)| proof.nibbles.starts_with(path))
+                    .map(|(_, node)| node.clone())
+                    .collect(),
+            );
```

Comment:
makes sense, although seeing the word `proof`/`proofs` everywhere made it sort of hard to read, not sure about better names